mor.c

```diff
@@ -0,0 +1,218 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <efi.h>
+#include <handler.h>
+#include <mor.h>
+
+#define MOR_CONTROL_LEN 1
+#define MOR_ACTION_VALID_MASK 0x11
+
+/* MemoryOverwriteRequestControl values */
+#define MOR_UNLOCKED 0
+#define MOR_LOCKED_WITHOUT_KEY 1
+#define MOR_LOCKED_WITH_KEY 2
+
+#define MOR_LOCK_REV1_UNLOCK 0
+#define MOR_LOCK_REV1_LOCK 1
+#define MOR_LOCK_REV1_LEN 1
+#define MOR_LOCK_REV2_LEN 8
+
+static uint8_t mor_key[MOR_LOCK_REV2_LEN];
+
+static const uint8_t MOR_CONTROL_NAME[] = {'M',0,'e',0,'m',0,'o',0,'r',0,'y',0,'O',0,'v',0,'e',0,'r',0,'w',0,'r',0,'i',0,'t',0,'e',0,'R',0,'e',0,'q',0,'u',0,'e',0,'s',0,'t',0,'C',0,'o',0,'n',0,'t',0,'r',0,'o',0,'l',0};
+static const uint8_t MOR_CONTROL_LOCK_NAME[] = {'M',0,'e',0,'m',0,'o',0,'r',0,'y',0,'O',0,'v',0,'e',0,'r',0,'w',0,'r',0,'i',0,'t',0,'e',0,'R',0,'e',0,'q',0,'u',0,'e',0,'s',0,'t',0,'C',0,'o',0,'n',0,'t',0,'r',0,'o',0,'l',0,'L',0,'o',0,'c',0,'k',0};
+const EFI_GUID morControlGuid =
+    {{0xbe, 0x39, 0x09, 0xe2, 0xd4, 0x32, 0xbe, 0x41, 0xa1, 0x50, 0x89, 0x7f, 0x85, 0xd4, 0x98, 0x29}};
+const EFI_GUID morControlLockGuid =
+    {{0xcf, 0x3c, 0x98, 0xbb, 0x1d, 0x15, 0xe1, 0x40, 0xa0, 0x7b, 0x4a, 0x17, 0xbe, 0x16, 0x82, 0x92}};
+
+bool
+setup_mor_variables(void)
+{
+    EFI_STATUS status;
+    uint8_t mor_control = 0, mor_control_lock = 0;
+    /*
+     * Initialize the MemoryOverwriteRequestControl variable to 0 at boot. The
+     * firmware never needs to clear the memory under Xen since bootup always
+     * takes place in a new domain with zeroed memory.
```

Comment:
Until we implement PPI... at which point do we need to wipe the PPI region?

Reply:
The PPI region is not necessarily real memory since all access goes via ACPI methods therefore I don't think it would be required by the spec to wipe it. Additionally, if it is backed by regular memory it will be marked as reserved so the OS wouldn't store secrets that could possibly be stolen in that area.

Furthermore, the save/restore of the PPI region should only be done for the small number of bytes needed to perform the command and save the result - the rest of the page should be left untouched (zeroed).